From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 22 Oct 2024 17:07:36 +0800
Subject: [PATCH] Cache blockstate cache array

Cache values to reduce allocations.
(See `Reduce array allocations` patch for full explanation)

The JMH benchmark of this patch can be found in SunBox's `CachedEnumValuesForLoop`

diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 2865473782fa9a31600b82d612048f0d1f2cec17..a9afca646255d4918da8697f73da738eec85813a 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -1135,6 +1135,10 @@ public abstract class BlockBehaviour implements FeatureElement {
         private static final class Cache {
             private static final Direction[] DIRECTIONS = Direction.values();
             private static final int SUPPORT_TYPE_COUNT = SupportType.values().length;
+            // Leaf start - Cache blockstate cache array
+            private static final SupportType[] SUPPORT_TYPE_VALUES = SupportType.values();
+            private static final Direction.Axis[] DIRECTION_AXIS_VALUES = Direction.Axis.values();
+            // Leaf end - Cache blockstate cache array
             public final VoxelShape collisionShape;
             public boolean largeCollisionShape;
             private final boolean[] faceSturdy;
@@ -1154,7 +1158,7 @@ public abstract class BlockBehaviour implements FeatureElement {
                 }
 
                 // Leaf start - Remove stream in BlockBehaviour cache blockstate
-                for (Direction.Axis axis : net.minecraft.core.Direction.Axis.VALUES) {
+                for (Direction.Axis axis : DIRECTION_AXIS_VALUES) { // Leaf - Cache blockstate cache array
                     if (this.collisionShape.min(axis) < 0.0 || this.collisionShape.max(axis) > 1.0) {
                         this.largeCollisionShape = true;
                         break;
@@ -1164,7 +1168,7 @@ public abstract class BlockBehaviour implements FeatureElement {
                 this.faceSturdy = new boolean[DIRECTIONS.length * SUPPORT_TYPE_COUNT];
 
                 for (Direction direction : DIRECTIONS) {
-                    for (SupportType type : SupportType.values()) {
+                    for (SupportType type : SUPPORT_TYPE_VALUES) { // Leaf - Cache blockstate cache array
                         this.faceSturdy[getFaceSupportIndex(direction, type)] = type.isSupporting(state, EmptyBlockGetter.INSTANCE, BlockPos.ZERO, direction);
                     }
                 }
